sdk/database/newdbplugin/database.go

```diff
@@ -0,0 +1,177 @@
+package newdbplugin
+
+import (
+	"context"
+	"time"
+)
+
+// Database to manipulate users within an external system (typically a database).
+type Database interface {
+	// Initialize the database plugin. This is the equivalent of a constructor for the
+	// database object itself.
+	Initialize(ctx context.Context, req InitializeRequest) (InitializeResponse, error)
+
+	// NewUser creates a new user within the database. This user is temporary in that it
+	// will exist until the TTL expires.
+	NewUser(ctx context.Context, req NewUserRequest) (NewUserResponse, error)
+
+	// UpdateUser updates an existing user within the database.
+	UpdateUser(ctx context.Context, req UpdateUserRequest) (UpdateUserResponse, error)
+
+	// DeleteUser from the database. This should not error if the user didn't
+	// exist prior to this call.
+	DeleteUser(ctx context.Context, req DeleteUserRequest) (DeleteUserResponse, error)
+
+	// Type returns the Name for the particular database backend implementation.
+	// This type name is usually set as a constant within the database backend
+	// implementation, e.g. "mysql" for the MySQL database backend. This is used
+	// for things like metrics and logging. No behavior is switched on this.
+	Type() (string, error)
+
+	// Close attempts to close the underlying database connection that was
+	// established by the backend.
+	Close() error
+}
+
+// ///////////////////////////////////////////////////////////////////////////
+// Database Request & Response Objects
+// These request and response objects are *not* protobuf types because gRPC does not
+// support all types that we need in a nice way. For instance, gRPC does not support
+// map[string]interface{}. It does have an `Any` type, but converting it to a map
+// requires extensive use of reflection and knowing what types to support ahead of
+// time. Instead these types are made as user-friendly as possible so the conversion
+// between protobuf types and request/response objects is handled by Vault developers
+// rather than needing to be handled by external plugin developers.
+// ///////////////////////////////////////////////////////////////////////////
+
+// ///////////////////////////////////////////////////////
+// Initialize()
+// ///////////////////////////////////////////////////////
+
+// InitializeRequest contains all information needed to initialize a database plugin.
+type InitializeRequest struct {
```

Comment:
Discussed this with Brian and even though this was the original direction that we wanted to go with, it might make sense to keep this middle layer around to keep the networking abstracted away from the interface in case we ever decide to move away from gRPC as the networking layer (like we previously did by going from net/rpc to gRPC). There's a bit more overhead by keeping these structs around, but since the conversion is mostly handled by us (the sdk), this shouldn't be an issue.